cpp/log/cert.cc

```diff
@@ -918,21 +902,35 @@ util::Status Cert::IsValidWildcardRedaction() const {
                         "No CT redaction count extension");
   }
 
-  // Unpack the extension contents, which should be SEQUENCE OF INTEGER
-  STACK_OF(ASN1_INTEGER)* const integers(
-      static_cast<STACK_OF(ASN1_INTEGER)*>(
-          ASN1_seq_unpack_ASN1_INTEGER(exty.ValueOrDie()->value->data,
-                                       exty.ValueOrDie()->value->length,
-                                       d2i_ASN1_INTEGER, ASN1_INTEGER_free)));
+  // Ensure the data in the extension is a sequence. DER encoding is same for
+  // SEQUENCE and SEQUENCE OF and we'll check types later.
+  if (exty.ValueOrDie()->value->data[0] !=
+      (V_ASN1_SEQUENCE | V_ASN1_CONSTRUCTED)) {
+    LOG(WARNING) << "CT redaction count extension is not a SEQUENCE OF";
+    return util::Status(Code::INVALID_ARGUMENT,
+                        "CT redaction count extension not a sequence");
+  }
+
+  // Unpack the extension contents, which should be SEQUENCE OF INTEGER.
+  // For compatibility we unpack any sequence and check integer type as we go.
+  // Don't pass the pointer from the extension directly as it gets incremented
+  // during parsing.
+  const unsigned char* sequence_data(
+      const_cast<const unsigned char*>(exty.ValueOrDie()->value->data));
+  STACK_OF(ASN1_TYPE)*
+      const asn1_types(static_cast<STACK_OF(ASN1_TYPE)*>(d2i_ASN1_SEQUENCE_ANY(
```

Comment:
clang_format and Eclipse formatter disagreed on this one. I've now redone it manually, is it OK?